### DISCUSSION TITLE
Weird filtering order behaviour when searching in settings

### DISCUSSION THREAD

**Discussion by psehr** [ppy/osu#30372] [Ideas]
### Type

Cosmetic

### Bug description

Found out an occurence of weird ordering when filtering settings using the search box. 

Even though I couldn't quite understand most of the related code for search filtering, I think part of this issue comes from the fact that the search results are not sorted by relative string distance, it looks like if multiple sections are returned, the order stays as default. I think the sections should also be ordered by similarity, not just filtered.

For example when I type "ski" I would expect the Skins section to come up first, but it's actually the Language section being the first element (because of all the language names finishing in "ski").
Even if it's minor, I figured there may be other places where this could be a problem so I wanted to report it.

### Screenshots or videos

![image](https://github.com/user-attachments/assets/a59e678f-cc6d-4e22-a4ba-3356476fcbd0)

### Version

2024.1009.1

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/17450341/compressed-logs.zip)

**Comment by bdach** [comment]
Converting to issue because to claim this is a 'clear bug' is a stretch. This is a feature request, and there is no clear way on how to proceed.

> I think the sections should also be ordered by similarity, not just filtered.

Reordering the sections would make them no longer match the breadcrumbs on the left.

If anything the 'most relevant' result could be scrolled into view automagically, but I am rather skeptical if it's possible to reliably automagically guess user intent.

**Comment by psehr** [comment]
That's a fair answer. Out of curiosity, as arbitrary as it is, couldn't some of the inner content of the dropdowns be ignored as search terms? Like the language one for example, since the list is quite long it comes up really often when typing 2-3 characters.

**Reply by Joehuu** [reply]
I think the solution is not to make "ski" or any letters that doesn't start at the beginning of the word match. We also have problems with mod searching (like "ns" brings up "transform" because it has "ns" somewhere in between).

**Reply by bdach** [reply]
> I think the solution is not to make "ski" or any letters that doesn't start at the beginning of the word match.

I tend to disagree with this. I'd prefer deprioritising the "wrong" results to just removing them entirely somehow.

> We also have problems with mod searching (like "ns" brings up "transform" because it has "ns" somewhere in between).

Irrelevant and will be addressed by https://github.com/ppy/osu/pull/30330.
